File: driver/src/error_print.rs
use frontend::{Diagnostic, lexer::Span};

use crate::cli::Input;
use termion::{color, style};

// if we see a file which is longer than 999'999 lines, we do not care anymore
const PREFIX_LEN: usize = 6;

const UNDERLINE_COLOR: color::Green = color::Green;

fn print_prefix(line: Option<usize>) {
    if let Some(line) = line {
        // print prefix that looks like `  42 | `
        eprint!("{:>PREFIX_LEN$} | ", line);
    } else {
        // print prefix that looks like `     | `
        eprint!("{:>PREFIX_LEN$} | ", "");
    }
}

fn print_overlapping_lines(input: &str, span: Span) {
    let lines_to_show = get_line_boundaries(input)
        .enumerate()
        .zip(input.lines())
        .skip_while(|&((_, (_, e)), _)| e < span.start)
        .take_while(|&((_, (b, e)), _)| b <= span.end && e >= span.start)
        .map(|((i, boundaries), line)| (i + 1, boundaries, line));

    // loop below is used to make pretty underlines
    // examples:
    //   37 | a := foobar
    //      |      ^~~~~^
    //
    //   42 | a := foo(1,
    //      |      ^~~~~~
    //   43 |          2,
    //      |          ~~
    //   44 |          3)
    //      |          ~^

    let mut i;
    for (number, (start_offset, end_offset), line) in lines_to_show {
        print_prefix(Some(number));
        eprintln!("{}", line);

        print_prefix(None);

        i = start_offset;
        while i < span.start {
            eprint!(" ");
            i += 1;
        }

        let mut started_underlining = false;
        if i == span.start {
            eprint!("{}^", color::Fg(UNDERLINE_COLOR));
            started_underlining = true;
            i += 1;
        }

        let mut line_chars = line.chars().skip(i - start_offset);
        while i < end_offset && i < span.end {
            // `unwrap_or` is needed because underline may be under a `\n`,
            // which is removed by `.lines()`
            if line_chars.next().map(char::is_whitespace).unwrap_or(false) && !started_underlining {
                eprint!(" ");
            } else {
                eprint!("{}~", color::Fg(UNDERLINE_COLOR));
                started_underlining = true;
            }
            i += 1;
        }

        // `span.start != span.end` to verify that we don't print "^^" under
        // a single character token
        if i == span.end && span.start != span.end {
            eprint!("{}^", color::Fg(UNDERLINE_COLOR));
        }
        eprintln!("{}", color::Fg(color::Reset));
    }
}

fn get_line_boundaries<'b>(source: &'b str) -> impl Iterator<Item = (usize, usize)> {
    source
        .chars()
        .enumerate()
        .filter_map(|(i, c)| if c == '\n' { Some(i) } else { None })
        .scan(0, |i: &mut usize, l: usize| {
            let res = (*i, l);
            *i = l + 1;
            Some(res)
        })
}

fn print_with_location(line: usize, column: usize, msg: &str) {
    eprintln!("{}:{}: {}", line, column, msg);
}

fn print_msg(input: &Input, span: Span, msg: &str) {
    let (line_number, column_number) = {
        let mut line_boundaries = get_line_boundaries(&input.contents).enumerate();

        let (number, (b, _)) = line_boundaries
            .find(|&(_, (lb, le))| lb <= span.end && le >= span.start)
            .expect("expected valid span");
        (number + 1, span.start - b + 1)
    };

    print_with_location(line_number, column_number, &msg);
    print_overlapping_lines(&input.contents, span);
}

pub fn print_error(input: &Input, diagnostic: &Diagnostic) {
    eprintln!("{}{}{}:", style::Bold, input.name, style::Reset);
    print_msg(
        &input,
        diagnostic.span,
        &format!(
            "{}{}error{}: {}{}",
            style::Bold,
            color::Fg(color::Red),
            color::Fg(color::Reset),
            diagnostic.msg,
            style::Reset,
        ),
    );

    for note in &diagnostic.notes {
        print_msg(
            &input,
            note.span,
            &format!(
                "{}{}note{}: {}{}",
                style::Bold,
                color::Fg(color::Cyan),
                color::Fg(color::Reset),
                note.value,
                style::Reset,
            ),
        );
    }
}

File: src/main/scala/driver/ErrorPrint.scala
package nop
package driver

import scala.io.AnsiColor.*
import frontend.lexer.Span

private val UNDERLINE_COLOR = GREEN

def printDiagnostic(filename: String, contents: String, diagnostic: Diagnostic) = {
  println(s"$BOLD$filename$RESET")
  printMsg(
    filename,
    contents,
    diagnostic.span,
    s"${BOLD + RED}error${RESET + BOLD}: ${diagnostic.msg}$RESET",
  )

  for note <- diagnostic.notes do
    printMsg(
      filename,
      contents,
      note.span,
      s"${BOLD + CYAN}note${RESET + BOLD}${note.value}$RESET",
    )
}

private def printMsg(filename: String, contents: String, span: Span, msg: String): Unit = {
  val lineNumber -> columnNumber = {
    val lineBoundaries = getLineBoundaries(contents).zipWithIndex

    val ((b, _), number) = lineBoundaries
      .find((boundaries, _) => boundaries._1 <= span.end && boundaries._2 >= span.start)
      .get
    (number + 1, span.start - b + 1)
  }

  printWithLocation(lineNumber, columnNumber, msg)
  printOverlappingLines(contents, span)
}

private def printWithLocation(lineNumber: Int, columnNumber: Int, msg: String): Unit =
  println(s"$lineNumber:$columnNumber: $msg")

private def printOverlappingLines(contents: String, span: Span): Unit = {
  def flatten(ctx: (((Int, Int), Int), String)): ((Int, Int), Int, String) =
    val ((boundaries, i), line) = ctx
    (boundaries, i, line)

  def isEarlyLine(ctx: ((Int, Int), Int, String)): Boolean =
    val ((_, e), _, _) = ctx
    e < span.start

  def isOverlappingLine(ctx: ((Int, Int), Int, String)): Boolean =
    val ((b, e), _, _) = ctx
    b <= span.end && e >= span.start

  val linesToShow = getLineBoundaries(contents).zipWithIndex
    .zip(contents.linesIterator)
    .map(flatten)
    .dropWhile(isEarlyLine)
    .takeWhile(isOverlappingLine)
    .map((boundaries, i, line) => (i + 1, boundaries, line))

  // loop below is used to make pretty underlines
  // examples:
  //   37 | a := foobar
  //      |      ^~~~~^
  //
  //   42 | a := foo(1,
  //      |      ^~~~~~
  //   43 |          2,
  //      |          ~~
  //   44 |          3)
  //      |          ~^

  var i: Int = 0
  for (number, (start_offset, end_offset), line) <- linesToShow do {
    printPrefix(number)
    println(line)

    printPrefix

    i = start_offset;
    while i < span.start do
      print(" ")
      i += 1

    var startedUnderlining = false;
    if i == span.start then
      print(s"$UNDERLINE_COLOR^")
      startedUnderlining = true
      i += 1

    var lineChars = line.iterator.drop(i - start_offset)
    while i < end_offset && i < span.end do {
      // `getOrElse` is needed because underline may be under a `\n`,
      // which is removed by `.linesIterator`
      if lineChars.nextOption
          .map(_.isWhitespace)
          .getOrElse(false) && !startedUnderlining
      then print(" ")
      else
        print(s"$UNDERLINE_COLOR~")
        startedUnderlining = true
      i += 1
    }

    // `span.start != span.end` to verify that we don't print "^^" under
    // a single character token
    if i == span.end && span.start != span.end then print(s"$UNDERLINE_COLOR^")
    println(RESET)
  }
}

private def getLineBoundaries(source: String): Iterator[(Int, Int)] =
  source.iterator.zipWithIndex
    .filter((c, i) => c == '\n')
    .map((c, i) => i)
    .scanLeft((0, -1)) { (prevRes, l) => (prevRes._2 + 1, l) }
    .drop(1)

private def printPrefix: Unit            = print(" " * 6 + " | ")
private def printPrefix(line: Int): Unit = print(f"$line%-6s | ")
